### PR TITLE
CDAP-6159 6158 Modify input class to support cross namespace stream/dataset access

### PR DESCRIPTION
**Pull request by lumig242**
- Change the API in Input class for datasetInput and streamInput.
- Modify the MapReduce program to make changes take effect.
  - To load stream from a different namespace, use `context.addInput(Input.ofStream("stream").fromNamespace("ns"));`

### CODE REVIEW

File: cdap-unit-test/src/test/java/co/cask/cdap/batch/stream/NoMapperApp.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public void initialize() throws Exception {
       MapReduceContext context = getContext();
       Job job = context.getHadoopJob();
       job.setReducerClass(NoMapperReducer.class);
-      context.addInput(Input.ofStream("nomapper"));
```
Comment (rohitsinha54) [inline cdap-unit-test/src/test/java/co/cask/cdap/batch/stream/NoMapperApp.java:55]: Dont get rid of no namespace test. keep that too.

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 2]
```diff
@@ -74,41 +74,38 @@ public Input alias(String alias) {
    *
    * @param datasetName the name of the input dataset
    */
-  public static Input ofDataset(String datasetName) {
+  public static DatasetInput ofDataset(String datasetName) {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java:77]: why this change?

Comment (lumig242) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java:77]: Users are expected to use the API as `Input.ofDataset().fromNamespace()`.
Without this change, a type conversion is required as `((DatasetInput) Input.ofDataset()).fromNamespace()`

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java:77]: got it

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 3]
```diff
@@ -206,10 +201,21 @@ private DatasetInput(String name, Map<String, String> arguments, @Nullable Itera
...
+    public String getNamespace() {
+      return namespace;
+    }
+
     @Nullable
     public List<Split> getSplits() {
       return splits;
     }
+
+    public <T extends Input> T fromNamespace(String namespace) {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java]: I don't think we need type parameterization here.
Can simply return DatasetInput.

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 4]
```diff
@@ -218,10 +224,22 @@ private DatasetInput(String name, Map<String, String> arguments, @Nullable Itera
...
     private final StreamBatchReadable streamBatchReadable;
+    private String namespace;
 
     private StreamInput(StreamBatchReadable streamBatchReadable) {
       super(streamBatchReadable.getStreamName());
       this.streamBatchReadable = streamBatchReadable;
+      this.namespace = null;
+    }
+
+    public <T extends Input> T fromNamespace(String namespace) {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java]: same here

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 5]
```diff
@@ -218,10 +224,22 @@ private DatasetInput(String name, Map<String, String> arguments, @Nullable Itera
   public static class StreamInput extends Input {
 
     private final StreamBatchReadable streamBatchReadable;
+    private String namespace;
 
     private StreamInput(StreamBatchReadable streamBatchReadable) {
       super(streamBatchReadable.getStreamName());
       this.streamBatchReadable = streamBatchReadable;
+      this.namespace = null;
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java]: We don't have this null initialization explicit in DatasetInput. We don't need it, right?

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 6]
```diff
@@ -206,10 +201,21 @@ private DatasetInput(String name, Map<String, String> arguments, @Nullable Itera
       return arguments;
     }
 
+    public String getNamespace() {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java:205]: mark this nullable

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java [hunk 7]
```diff
@@ -218,10 +224,22 @@ private DatasetInput(String name, Map<String, String> arguments, @Nullable Itera
...
+      this.namespace = null;
+    }
+
+    public <T extends Input> T fromNamespace(String namespace) {
+      StreamInput streamInput = new StreamInput(streamBatchReadable);
+      streamInput.namespace = namespace;
+      return (T) streamInput;
+    }
+
+    public String getNamespace() {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java:242]: mark this nullable

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java (hunk 5 above)
Comment (lumig242) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java]: Yes, we don't need it. Removed.

**Comment by lumig242** [comment]
Addressed the comments above.

**Comment by anew** [comment]
change in the last commit looks good. Is the same change needed for Spark?

**Comment by lumig242** [comment]
I haven't look into spark now but I think some similar changes will be needed for Spark. I'm almost done with the change in Flowlet programs, will check Spark after this.

**Comment by anew** [comment]
LGTM

**Comment by rohitsinha54** [comment]
Please add build link to PR. Do we have a successful build for this ?

**Comment by lumig242** [comment]
The build link: http://builds.cask.co/browse/CDAP-DUT4295-4/test
